runtime/src/account_info.rs

```diff
@@ -63,22 +63,25 @@ pub struct AccountInfo {
 
     /// needed to track shrink candidacy in bytes. Used to update the number
     /// of alive bytes in an AppendVec as newer slots purge outdated entries
-    /// Note that highest bit is used for ZERO_LAMPORT_BIT
-    stored_size: StoredSize,
+    /// Note that highest bits are used by ALL_BITS
+    /// So, 'stored_size' is 'stored_size_mask' with ALL_BITS masked out.
+    stored_size_mask: StoredSize,
 }
 
-/// presence of this bit in stored_size indicates this account info references an account with zero lamports
+/// These bits can be present in stored_size_mask to indicate additional info about the AccountInfo
 const ZERO_LAMPORT_BIT: StoredSize = 1 << (StoredSize::BITS - 1);
+const CACHED_STORE_ID_BIT: StoredSize = 1 << (StoredSize::BITS - 2);
+const ALL_BITS: StoredSize = ZERO_LAMPORT_BIT | CACHED_STORE_ID_BIT;
 
 impl ZeroLamport for AccountInfo {
     fn is_zero_lamport(&self) -> bool {
-        self.stored_size & ZERO_LAMPORT_BIT == ZERO_LAMPORT_BIT
+        self.stored_size_mask & ZERO_LAMPORT_BIT == ZERO_LAMPORT_BIT
     }
 }
 
 impl IsCached for AccountInfo {
     fn is_cached(&self) -> bool {
-        self.store_id == CACHE_VIRTUAL_STORAGE_ID
+        self.stored_size_mask & CACHED_STORE_ID_BIT == CACHED_STORE_ID_BIT
```

Comment:
@brooksprumo extending the concern from here: https://github.com/solana-labs/solana/pull/21842/files#r769215769

It seems like we'll need a migration to maintain compatibility with old snapshots where the `AppendVecId` was still a u64. 

Also wanted to make sure that the AppendVec remapping logic on snapshot restart doesn't accidentally write this `CACHED_STORE_ID_BIT` on startup. We should have about 4 billion appendvecs in a u32, but this bit is `1 << (StoredSize::BITS - 2);` so we have 4x less than that, about 500 million appendvecs. That's only about 8 years?

Reply:
I think the type used for the snapshots w.r.t. the AppendVec IDs is staying the same, right @jeffwashington? It's just our in-mem representation. Also, since the bits are being stolen from `store_size` instead of the `store_id`, we will still have the full 2^32 append vec IDs. I think that'll work out OK, yes?

Reply:
ah shoot for some reason in my mind I believed `AppendVec` ids were u64s, yeah the deserialization shouldn't be an issue then

Reply:
The intention is for the serialized type in snapshots to remain as it was. So, no migration should be needed.

Reply:
We previously removed 1 bit from the u64 for the special 'cached' state. We asserted if that bit was ever used for real append vecs.
Now, with this PR, we do not reserve a bit from the u64 id for the special 'cached' state. Thus, all possible values for AppendVecId are valid append vec ids. Changing the type of AppendVecId to u32 is in the follow-on PR. We do need to make sure the id wrap case does work correctly. As well as the load from snapshot case.